certbot/plugins/webroot.py

```diff
@@ -207,7 +208,7 @@ def _perform_single(self, achall):
         old_umask = os.umask(0o022)
 
         try:
-            with open(validation_path, "wb") as validation_file:
+            with safe_open(validation_path, mode="wb", chmod=0o644) as validation_file:
```

Comment:
I see a couple changes from `open` to `safe_open` in this PR; what was your decision-making process for which ones to change and which ones to leave?

Reply:
In fact, the builtin method `open` is not creating files that are correctly set up in terms of DACL and owner for Windows. I changed from builtin `open` to `certbot.util.safe_open` each time a test related to the concerned logic was asserting things about the security of its content (and basically was failing without `safe_open`).

Reply:
Well, glad the tests were catching it! Makes me wonder, if the tests just caught those cases, maybe we're just lacking tests for other cases but should be doing the same thing. Do we need to disallow builtin `open`?

Reply:
As a result of the approach described in my previous comment, there is effectively not call to builtin `open` or `os.open` in certbot production code anymore. It may be interesting to lock this out using a pylint rule. I will check what is possible here.

Reply:
There is a lot of use of `open` in Certbot codebase. I agree with you that we should propose a complete replacement for `open` in Certbot in the future, but I think also it is not necessary now.

Indeed, these calls to `open` are not setting any permission on Linux, so one can expect that it is not a problem also to not control the permissions on Windows. It would be an overfit for Windows compared to Linux, and would impact significantly Certbot behavior on Linux.

Here `safe_open` and `os.open` are used in Certbot explicitly to open files in a safe way (in term of concurrency and in term of permission). For everywhere it is relevant, theses methods are used with a proper handle through `filesystem` module, and there are tested.

Reply:
> Indeed, these calls to open are not setting any permission on Linux, so one can expect that it is not a problem also to not control the permissions on Windows.

This can't be quite true -- if a file is created, it has to be created with *some* permissions.

The thing that surprises me is that there are calls to `open` that turn out to need `safe_open`. I would have thought that anything that needs `safe_open` would have already been using `os.open`. For the ones that were changed here, do you already know why it would have been fine on linux before? If not I'll check myself, I don't want to be duplicating work you may have already done.

Reply:
A file has obviously always a set of permissions defined. My sentence here had to be interpreted as "these calls to open are not setting any _specific_ permission".

So for the calls to `open` that was fine on Linux, but not on Windows:
* in `log.py`, it was because `NamedTemporaryFile`, despite the fact that it is claimed to create a temporary file in the most secured manner, is not on Windows. The bare call to `os.open` just let the DACL be inherited from upper level. On Linux, mode is set to 0o700 which is fine.
* in `webroot.py`, it was because `os.umask` is invoked before creating the file. So on Linux this umask is honored and file is created with the correct permissions. But Windows is not using this umask, so we need to use the `filesystem` functions to make this work correctly.

The last situation I anticipated, but it is not encountered in certbot codebase, would be a call to `os.open` with `mode` parameter explicitly set.